### ISSUE TITLE
System.ComponentModel.EventBasedAsync.AsyncOperationTests.Cancel Failure in CI Build

### ISSUE THREAD

**Issue by nslottow** [dotnet/corefx#666]
I just ran into this failure that seems like a race condition:
```
Starting:    System.ComponentModel.EventBasedAsync.Tests
     System.ComponentModel.EventBasedAsync.AsyncOperationTests.Cancel [FAIL]
        System.AggregateException : One or more errors occurred.
        ---- Assert.True() Failure
        Stack Trace:
              at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
              at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
              at System.Threading.Tasks.Task.Wait()
           d:\j\workspace\dotnet_corefx_prtest\Config\Debug\OS\Windows_NT\Tests\false\label\windows\src\System.ComponentModel.EventBasedAsync\tests\AsyncOperationTests.cs(65,0): at System.ComponentModel.EventBasedAsync.AsyncOperationTests.Cancel()
           ----- Inner Stack Trace -----
           d:\j\workspace\dotnet_corefx_prtest\Config\Debug\OS\Windows_NT\Tests\false\label\windows\src\System.ComponentModel.EventBasedAsync\tests\AsyncOperationTests.cs(77,0): at System.ComponentModel.EventBasedAsync.AsyncOperationTests.<Cancel>b__18()
              at System.Threading.Tasks.Task.InnerInvoke()
              at System.Threading.Tasks.Task.Execute()
```
I haven't been able to repro this locally.  It may be worth trying on a single core machine.

**Comment by stephentoub** [comment]
Just eye-balling the code, I believe the race stems from lines 209-211.  The code is setting the completed event and then setting the canceled event.  As soon as completed is set, the wait on line 153 and thus line 75 can wake up.  It's then a race as to whether line 211 sets canceled before line 77 and thus line 131 checks whether canceled is set.

**Comment by nslottow** [comment]
That makes sense.  So as long as _cancelEvent is set before _completeEvent in OnOperationCompleted(), the wait on line 75 won't be awoken prematurely with a chance of reading operation.Cancelled before it is set.  I'll send a PR to fix this.

**Comment by nslottow** [comment]
Closed by #677
